processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,496 @@
+      }
+
+      return AggregateResult.ok();
+    }
+    catch (RuntimeException e) {
+      finished = true;
+      throw e;
+    }
+  }
+
+  /**
+   * Checks two keys contained in the given buffers are same.
+   *
+   * @param curKeyBuffer the buffer for the given key from {@link #aggregate(Object)}
+   * @param buffer       the whole array buffer
+   * @param bufferOffset the key offset of the buffer
+   *
+   * @return true if the two buffers are same.
+   */
+  private boolean keyEquals(ByteBuffer curKeyBuffer, ByteBuffer buffer, int bufferOffset)
+  {
+    // Since this method is frequently called per each input row, the compare performance matters.
+    int i = 0;
+    for (; i + Long.BYTES <= keySize; i += Long.BYTES) {
+      if (curKeyBuffer.getLong(i) != buffer.getLong(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    if (i + Integer.BYTES <= keySize) {
+      // This can be called at most once because we already compared using getLong() in the above.
+      if (curKeyBuffer.getInt(i) != buffer.getInt(bufferOffset + i)) {
+        return false;
+      }
+      i += Integer.BYTES;
+    }
+
+    for (; i < keySize; i++) {
+      if (curKeyBuffer.get(i) != buffer.get(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Initialize a new slot for a new grouping key.  This may be potentially blocked if the array is full until at least
+   * one slot becomes available.
+   */
+  private void initNewSlot(ByteBuffer newKey)
+  {
+    // Wait if the array is full and increase curWriteIndex
+    increaseWriteIndex();
+
+    final int recordOffset = recordSize * curWriteIndex;
+    buffer.position(recordOffset);
+    buffer.put(newKey);
+
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(buffer, recordOffset + aggregatorOffsets[i]);
+    }
+  }
+
+  /**
+   * Wait for {@link #nextReadIndex} to be moved if necessary and move {@link #curWriteIndex}.
+   */
+  private void increaseWriteIndex()
+  {
+    // In the below, we check that the array is full and wait for at least one slot to become available.
+    //
+    // nextReadIndex is a volatile variable and the changes on it are continuously checked until they are seen in
+    // waitFor(). See the following links.
+    // * http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.3.1.4
+    // * http://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5
+    // * https://stackoverflow.com/questions/11761552/detailed-semantics-of-volatile-regarding-timeliness-of-visibility
+
+    if (curWriteIndex == maxNumSlots - 1) {
+      // We additionally check that nextReadIndex is -1 here because the writing thread should wait for the reading
+      // thread to start reading only when the writing thread tries to overwrite the first slot for the first time.
+      waitFor(
+          notUsed -> (nextReadIndex == -1 || nextReadIndex == 0) && !Thread.currentThread().isInterrupted(),
```

Comment:
`!Thread.currentThread().isInterrupted()` is a part of all conditions, better moved to `waitFor()` method

Reply:
Are you sure that a new lambda object is not created on each call? But if a lambda is cached in a field, it will harm readability

Reply:
Good point. I could see some performance degradation with using lambda. Reverted this change.